src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -3122,6 +3133,71 @@ NormalizeMultiRowInsertTargetList(Query *query)
 }
 
 
+/*
+ * GetNumberOfUnspecifiedDefaultCols returns number of default columns that
+ * are not explicitly specified in given INSERT query's values_lists. We do
+ * that by counting non-Var nodes in targetList of given query object.
+ */
+static int
+GetNumberOfUnspecifiedDefaultCols(Query *query)
+{
+	int unspecifiedDefaultCols = 0;
+
+	TargetEntry *targetEntry = NULL;
+	foreach_ptr(targetEntry, query->targetList)
+	{
+		Expr *targetExpr = targetEntry->expr;
+		if (!IsA(targetExpr, Var))
+		{
+			unspecifiedDefaultCols++;
+		}
+	}
+
+	return unspecifiedDefaultCols;
+}
+
+
+/*
+ * AppendUnspecifiedDefaultColReferences appends dummy reference names for each
+ * column that has a DEFAULT definition but not specified in VALUES clause of
+ * INSERT query.
+ */
+static void
+AppendUnspecifiedDefaultColReferences(RangeTblEntry *valuesRTE,
+									  int unspecifiedDefaultCols)
+{
+	/*
+	 * Postgres starts from 1 to name columns in VALUES clause. So we start
+	 * with next integer to naming columns that we explicitly append to target
+	 * list in NormalizeMultiRowInsertTargetList function.
+	 * Also see addRangeTableEntryForValues.
+	 */
+	int existingColReferences = list_length(valuesRTE->eref->colnames);
+	int unspecifiedColId = existingColReferences + 1;
+	for (int i = 0; i < unspecifiedDefaultCols; i++)
```

Comment:
adding always to the end seems dangerous. How does this work if the DEFAULT value is in between of two columns?

Reply:
As far as manually tested, it still works as expected. I can add a test for that

Reply:
I think we should try to have a better understanding than `tested and worked`. Can you debug and see how does that work? 

Also, on `NormalizeTargetList` function, we are already dealing with  default columns with the comment `/* copy the column's default expression */`.

Can't we handle what we do here there?

Reply:
>I think we should try to have a better understanding than tested and worked. Can you debug and see how does that work?

I had put some breakpoints and checked how postgres normally processes `values` list.
If we insert into a table from a values list, postgres just creates a list of dummy `String` objects for such columns like `column1`, `column2`, no matter which columns we are inserting into or in which order.

However, I will look into that once more.

-----

>Also, on NormalizeTargetList function, we are already dealing with default columns with the comment /* copy the column's default expression */.
Can't we handle what we do here there?

We can actually, but I didn't want to complicate the stuff that the below loop does.

Reply:
>However, I will look into that once more.

Debugged a bit more and also looked into pretty-printed query tree once more.
Compared two queries:


1- INSERT INTO reference_table(d, a, b, c) VALUES (40,40,40,40), (50,50,50,50);
2- INSERT INTO reference_table(c, a) VALUES (40,40), (50,50);

On master, we had two differences in pretty-printed query tree (excluding `location` fields in `Const` expressions):

1- `valuesRTE->eref->colnames` differs (which agrees my debugging):
```diff
       {RTE 
       :alias <> 
       :eref 
          {ALIAS 
          :aliasname *VALUES* 
-         :colnames ("column1" "column2" "column3" "column4")
+         :colnames ("column1" "column2")
          }
```

2- 

```diff
    :rtable (
       {RTE 
       :alias <> 
       :eref 
          {ALIAS 
          :aliasname citus_local_table 
          :colnames ("a" "b" "c" "d")
          }
       :rtekind 0 
       :relid 17711 
       :relkind r 
       :rellockmode 3 
       :tablesample <> 
       :lateral false 
       :inh false 
       :inFromCl false 
       :requiredPerms 1 
       :checkAsUser 0 
       :selectedCols (b)
-      :insertedCols (b 8 9 10 11)
+      :insertedCols (b 8 10)
```

I think skipping second diff is already okay as I think we shouldn't check for permissions additionally for default columns (as postgres normally doesn't do that).

---

The reason that resolving the first difference fixes the issue is that:

We were crashing (or assertion failuring) in `add_vars_to_targetlist` because when we don't add such dummy column references, postgres doesn't set `attr_needed` & `min/max_attr` fields of `RelOptInfo` struct for `valuesRTE` properly, which is processed in  `add_vars_to_targetlist`

That `RelOptInfo` is actually built in `build_simple_rel` and that explains why adding dummy column references fixes the issue, from `relnode.c:322`:
```c
			rel->min_attr = 0;
			rel->max_attr = list_length(rte->eref->colnames);
			rel->attr_needed = (Relids *)
				palloc0((rel->max_attr - rel->min_attr + 1) * sizeof(Relids));
			rel->attr_widths = (int32 *)
				palloc0((rel->max_attr - rel->min_attr + 1) * sizeof(int32));
			break;
```